qutip/hsolve.py

```diff
@@ -0,0 +1,306 @@
+from qutip import spre, spost, sprepost, thermal_dm, mesolve, Options, dims
+from qutip import tensor, identity, destroy, sigmax, sigmaz, basis, qeye
+from qutip import liouvillian, mat2vec, state_number_enumerate
+
+def _heom_state_dictionaries(dims, excitations):
+    """
+    Return the number of states, and lookup-dictionaries for translating
+    a state tuple to a state index, and vice versa, for a system with a given
+    number of components and maximum number of excitations.
+    Parameters
+    ----------
+    dims: list
+        A list with the number of states in each sub-system.
+    excitations : integer
+        The maximum numbers of dimension
+    Returns
+    -------
+    nstates, state2idx, idx2state: integer, dict, dict
+        The number of states `nstates`, a dictionary for looking up state
+        indices from a state tuple, and a dictionary for looking up state
+        state tuples from state indices.
+    """
+    nstates = 0
+    state2idx = {}
+    idx2state = {}
+
+    for state in state_number_enumerate(dims, excitations):
+        state2idx[state] = nstates
+        idx2state[nstates] = state
+        nstates += 1
+
+    return nstates, state2idx, idx2state
+
+
+def _heom_number_enumerate(dims, excitations=None, state=None, idx=0):
+    """
+    An iterator that enumerate all the state number arrays (quantum numbers on
+    the form [n1, n2, n3, ...]) for a system with dimensions given by dims.
+    Example:
+        >>> for state in state_number_enumerate([2,2]):
+        >>>     print(state)
+        [ 0.  0.]
+        [ 0.  1.]
+        [ 1.  0.]
+        [ 1.  1.]
+    Parameters
+    ----------
+    dims : list or array
+        The quantum state dimensions array, as it would appear in a Qobj.
+    state : list
+        Current state in the iteration. Used internally.
+    excitations : integer (None)
+        Restrict state space to states with excitation numbers below or
+        equal to this value.
+    idx : integer
+        Current index in the iteration. Used internally.
+    Returns
+    -------
+    state_number : list
+        Successive state number arrays that can be used in loops and other
+        iterations, using standard state enumeration *by definition*.
+    """
+
+    if state is None:
+        state = np.zeros(len(dims))
+
+    if excitations and sum(state[0:idx]) > excitations:
+        pass
+    elif idx == len(dims):
+        if excitations is None:
+            yield np.array(state)
+        else:
+            yield tuple(state)
+            
+    else:
+        for n in range(dims[idx]):
+            state[idx] = n
+            for s in state_number_enumerate(dims, excitations, state, idx + 1):
+                yield s
+
+
+def cot(x):
+    """
+    Calculate cotangent.
+    Parameters
+    ----------
+    x: Float
+        Angle.
+    """
+    return np.cos(x)/np.sin(x)
+
+
+def hsolve(H, Q,gam, lam0, Nc, N, Temperature, tlist, initial_state,
```

Comment:
use an argument order that is consistent with `mesolve`. rename initial_state to rho0 or psi0 (whichever is more suitable)